### ISSUE TITLE
Build fails if gmtime_s() or gmtime_r() are not available

### ISSUE THREAD

**Issue by RonEld** [ARMmbed/mbedtls#1907]
### Description
- Type: Bug
- Priority:  Major 

---------------------------------------------------------------
## Bug

**OS**  
Mbed OS
**mbed TLS build:**  
Version: 2.12.0
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):   ARM Compiler
Additional environment information:  

**Expected behavior**   
build to succeed
**Actual behavior**  
build failure:
~~~
"/tmp/H7cWpV", line 124 (column 3): Warning: L6312W: Empty Execution region description for region RW_IRAM1
Error: L6218E: Undefined symbol gmtime_r (referred from BUILD/K64F/ARM/mbed-os/features/mbedtls/src/x509.o).
Finished: 0 information, 1 warning and 1 error messages.
[ERROR] "/tmp/H7cWpV", line 124 (column 3): Warning: L6312W: Empty Execution region description for region RW_IRAM1
Error: L6218E: Undefined symbol gmtime_r (referred from BUILD/K64F/ARM/mbed-os/features/mbedtls/src/x509.o).
~~~
**Steps to reproduce**  
build an mbed OS application, with Mbed TLS 2.12.0, using mbed-cli

This is from change introduced in https://github.com/ARMmbed/mbedtls/commit/512b4ee9c7421c4d70352d2a37a6ef1038a515b0

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2436

**Comment by gilles-peskine-arm** [comment]
That's not a problem with armcc, it's a problem with libc compatibility. If the toolchain makes a difference, it's because it's picking up a different libc. `gmtime_r` is a POSIX function, not a standard C function. It's legitimate for a libc to have `gmtime` but not `gmtime_r`. We should use `gmtime_r` where it's present and `gmtime` otherwise.

**Comment by gilles-peskine-arm** [comment]
The documentation for `MBEDTLS_HAVE_TIME_DATE` states:

> System has time.h and time(), gmtime() and the clock is correct.

#1198 added the requirement that the platform must have `gmtime_r` or `gmtime_s`, but didn't change the documentation. I do think we should support `MBEDTLS_HAVE_TIME_DATE` on platforms with `gmtime` only. Mbed OS is a pretty compelling use case.

**Comment by andresag01** [comment]
To expand on the information above:

* This problem occurs with Mbed TLS version 2.12.0 and the current Mbed OS master (1023280bd9d8880239dd868f94a12fb0076d9028)
* As @gilles-peskine-arm, this is a problem with the libc used to build Mbed OS and Mbed TLS as gmtime_r is a POSIX function
* I have tested and can confirm that the problem happens with the Mbed OS cli toolchain options uARM and ARM. I suspect the problem also happens with IAR and ARMC6, but I have not tested this.
* From the Mbed TLS side, the issue was introduced in #1198, and from the Mbed OS side the issue was introduced in https://github.com/ARMmbed/mbed-os/pull/4846 which enables MBEDTLS_HAVE_TIME_DATE

@gilles-peskine-arm: I agree that the problem can be solved by detecting whether the libc is POSIX (by checking _POSIX_VERSION for example) and then using the appropriate function. However, as pointed out in #1198, the issue is that gmtime() is not thread safe. Do you think we should provide some sort of _ALT option, perhaps something else, to allow platforms like Mbed OS to deal with this sort of problem safely?

**Comment by sbutcher-arm** [comment]
There's actually an underlying problem here that armcc doesn't actually support `gmtime()` either! It actually [always returns NULL](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0378f/chr1358938934656.html).

The fix in #1927 will doubtless help someone, but it doesn't help Mbed OS. The PR in #1927 will now allow the code to build - but it still won't yield a useful date and time nor do the right certificate date checks.

Therefore, I've changed the title to 'Build fails if gmtime_s() or gmtime_r() are not available' and created a new bug of #1975 - 'Arm C Compiler doesn't support gmtime()'.

**Comment by andresag01** [comment]
@sbutcher-arm: Thanks for the information regarding the state of gmtime() for armcc.

While the PR #1927 does not fully resolve the armcc situation for Mbed OS, I do think that it is a first step towards fixing the issue for Mbed OS. This is because #1927 allows us to configure an implementation for gmtime() at compile time. So in the case of Mbed OS, we could detect the supported toolchains/libcs that do not have a gmtime() and automatically configure a suitable implementation.
